tesseract_process_managers/src/core/task_info.cpp

```diff
@@ -30,6 +30,46 @@ namespace tesseract_planning
 {
 TaskInfo::TaskInfo(std::size_t unique_id, std::string name) : unique_id(unique_id), task_name(std::move(name)) {}
 
+bool TaskInfo::operator==(const TaskInfo& rhs) const
+{
+  static auto max_diff = static_cast<double>(std::numeric_limits<float>::epsilon());
+
+  bool equal = true;
+  equal &= return_value == rhs.return_value;
+  equal &= unique_id == rhs.unique_id;
+  equal &= task_name == rhs.task_name;
+  equal &= message == rhs.message;
+  equal &= tesseract_common::almostEqualRelativeAndAbs(elapsed_time, rhs.elapsed_time, max_diff);
+  equal &= instructions_input == rhs.instructions_input;
+  equal &= instructions_output == rhs.instructions_output;
+  equal &= results_input == rhs.results_input;
+  equal &= results_output == rhs.results_output;
+  return equal;
+}
+bool TaskInfo::operator!=(const TaskInfo& rhs) const { return !operator==(rhs); }
+
+template <class Archive>
+void TaskInfo::serialize(Archive& ar, const unsigned int /*version*/)
+{
+  ar& boost::serialization::make_nvp("return_value", return_value);
+  ar& boost::serialization::make_nvp("unique_id", unique_id);
+  ar& boost::serialization::make_nvp("task_name", task_name);
+  ar& boost::serialization::make_nvp("message", message);
+  ar& boost::serialization::make_nvp("elapsed_time", elapsed_time);
+
+  ar& boost::serialization::make_nvp("instructions_input", instructions_input);
+  ar& boost::serialization::make_nvp("instructions_output", instructions_output);
+  ar& boost::serialization::make_nvp("results_input", results_input);
+  ar& boost::serialization::make_nvp("results_output", results_output);
+
+  //  ar& boost::serialization::make_nvp("environment", environment);
+}
+
+#include <boost/archive/xml_oarchive.hpp>
+#include <boost/archive/xml_iarchive.hpp>
+template void TaskInfo::serialize(boost::archive::xml_oarchive& ar, const unsigned int version);
+template void TaskInfo::serialize(boost::archive::xml_iarchive& ar, const unsigned int version);
```

Comment:
Recommend moving this to the end outside the namespace

``` c++
#include <boost/archive/xml_oarchive.hpp>
#include <boost/archive/xml_iarchive.hpp>
template void tesseract_planning::TaskInfo::serialize(boost::archive::xml_oarchive& ar, const unsigned int version);
template void tesseract_planning::TaskInfo::serialize(boost::archive::xml_iarchive& ar, const unsigned int version);
```

Reply:
Let me know if this solves the serialization issues.

Reply:
That doesn't seem to have fixed it.

Reply:
@Levi-Armstrong I added instructions to the serialization test, and that seems to work. However, it still doesn't work in my application. Maybe there is a bug in the saveInputs function? Or I'm doing something wrong?

```
    auto info_map = response->interface->getTaskInfoMap();
    std::vector<TaskInfo> task_info_vec;
    for (const auto& info : info_map)
      task_info_vec.push_back(*info.second);
    std::string task_info_string = Serialization::toArchiveStringXML<std::vector<TaskInfo>>(task_info_vec);
```

Reply:
Have you confirmed that the derived class stored as the base class still serialized ok?

Reply:
It looks like you are serializing a `std::vector` but did you include `#include <boost/serialization/vector.hpp>` in the file?

Reply:
Yeah I did. The vector serializes fine. Serializing the base class seems to work. This passes and the xml file has instructions in it.

```
TEST(TesseractProcessManagersSerializeUnit, serializeDiscreteContactCheckTaskInfo)  // NOLINT
{
  auto task_info = std::make_shared<DiscreteContactCheckTaskInfo>(123456, "my task");
  task_info->elapsed_time = 123.456;
  task_info->message = "Test message";
  task_info->instructions_input = rasterExampleProgram();
  task_info->instructions_output = freespaceExampleProgramABB();
  task_info->results_input = freespaceExampleProgramIIWA();
  task_info->results_output = rasterDTExampleProgram();

  {  // Archive program to file
    std::string file_path = tesseract_common::getTempPath() + "serializeDiscreteContactCheckTaskInfo.xml";
    EXPECT_TRUE(Serialization::toArchiveFileXML<TaskInfo>(*task_info, file_path));

    TaskInfo ntask_info = Serialization::fromArchiveFileXML<TaskInfo>(file_path);
    EXPECT_TRUE(*task_info == ntask_info);
  }

  {  // Archive program to string
    std::string task_info_string = Serialization::toArchiveStringXML<TaskInfo>(*task_info, "task_info");
    EXPECT_FALSE(task_info_string.empty());

    TaskInfo ntask_info = Serialization::fromArchiveStringXML<TaskInfo>(task_info_string);
    EXPECT_TRUE(*task_info == ntask_info);
  }
}

```

Reply:
@Levi-Armstrong I just realized that I was being stupid. There is a flag to turn on/off saving the Instructions and it defaults to off. We don't currently have a good way of setting that in the planning server, but regardless I think this is fine for now.